### PR TITLE
add registry column to image table

### PR DESCRIPTION
**Pull request by NatalieEO**
https://github.com/dockstore/dockstore/issues/3275

### CODE REVIEW

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/Image.java [hunk 1]
```diff
@@ -61,6 +61,10 @@
     @ApiModelProperty(value = "Docker ID of the image", position = 4)
     private String imageID;
 
+    @Column()
+    @ApiModelProperty(value = "Registry image belongs to", position = 5)
```
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/core/Image.java]: Real nit-picky, but maybe `Registry the image belongs to`? At first glance I read `Registry image` as noun phrase

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/AbstractImageRegistry.java [hunk 2]
```diff
@@ -330,7 +330,7 @@ private void logToolRefresh(final String dashboardPrefix, final Tool tool) {
                         final String manifestDigest = i.getDigest();
                         checksums.add(new Checksum(manifestDigest.split(":")[0], manifestDigest.split(":")[1]));
                     }
-                    Set<Image> images = Collections.singleton(new Image(checksums, repo, tag.getName(), r.getImageID()));
+                    Set<Image> images = Collections.singleton(new Image(checksums, repo, tag.getName(), r.getImageID(), tool.getRegistry()));
                     tag.getImages().addAll(images);
```
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/AbstractImageRegistry.java]: You don't need to create an intermediate Set; you can just add the image directly by calling `tag.getImages().add(<image>)`.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/AbstractImageRegistry.java [hunk 3]
```diff
@@ -595,7 +595,7 @@ private void updateImageInformation(Tool tool, Tag newTag, Tag oldTag) {
                                 List<Checksum> checksums = new ArrayList<>();
 
                                 checksums.add(new Checksum(manifestDigest.split(":")[0], manifestDigest.split(":")[1]));
-                                Set<Image> images = Collections.singleton(new Image(checksums, tool.getNamespace() + '/' + tool.getName(), tag.getName(), null));
+                                Set<Image> images = Collections.singleton(new Image(checksums, tool.getNamespace() + '/' + tool.getName(), tag.getName(), null, tool.getRegistry()));
```
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/AbstractImageRegistry.java]: Same, you don't need intermediate set

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/QuayImageRegistry.java [hunk 4]
```diff
@@ -121,7 +121,7 @@ public QuayImageRegistry(final Token quayToken) {
             final String imageID = quayTag.getImageId();
             List<Checksum> checksums = new ArrayList<>();
             checksums.add(new Checksum(manifestDigest.split(":")[0], manifestDigest.split(":")[1]));
-            return Collections.singleton(new Image(checksums, repo, tag.getName(), imageID));
+            return Collections.singleton(new Image(checksums, repo, tag.getName(), imageID, tool.getRegistry()));
```
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/QuayImageRegistry.java]: Following up on my previous comments, if this always returns a max of one Image (?), then the method should return an Optional<Image> and not a Set.

File: dockstore-integration-testing/src/test/java/io/dockstore/client/cli/WorkflowIT.java [hunk 5]
```diff
@@ -889,6 +890,10 @@ private void testTRSConversion(final List<ToolVersion> versions, final String sn
                 assertTrue(trsVersion.isIsProduction());
                 assertEquals("There should be" + numImages + "image(s) in this workflow", numImages, trsVersion.getImages().size());
                 snapshotInList = true;
+                for (ImageData imageData :trsVersion.getImages()) {
```
Comment (garyluu) [inline dockstore-integration-testing/src/test/java/io/dockstore/client/cli/WorkflowIT.java]: Assert getImages() not empty.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/Image.java [hunk 6]
```diff
@@ -61,6 +61,10 @@
...
+    private String imageRegistry;
```
Comment (denis-yuen) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/core/Image.java]: Should this just use an Enum?
See Organization.status for an example. 
The enum would seem to just be io.dockstore.common.Registry

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- dockstore-common/src/main/java/io/dockstore/common/Registry.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dockstore/dockstore/pullRequest?prid=5105070&bid=16789291)

File: dockstore-webservice/src/main/resources/swagger.yaml [hunk 7]
```diff
@@ -6874,6 +6874,16 @@ definitions:
         type: "string"
         position: 4
         description: "Docker ID of the image"
+      imageRegistry:
+        type: "string"
+        position: 5
+        description: "Registry the image belongs to"
+        enum:
+        - "QUAY_IO"
```
Comment (garyluu) [inline dockstore-webservice/src/main/resources/swagger.yaml]: Swagger doesn't match openapi

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/QuayImageRegistry.java [hunk 8]
```diff
@@ -112,7 +113,7 @@ public QuayImageRegistry(final Token quayToken) {
     }
 
     //TODO: If the repo has a lot of tags, then it needs to be paged through. Can get tag info individually, but then that's more API calls.
-    private Set<Image> getImagesForTag(final Tool tool, final Tag tag, final QuayTag quayTag) {
+    private Optional<Image> getImageForTag(final Tool tool, final Tag tag, final QuayTag quayTag) {
```
Comment (denis-yuen) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/QuayImageRegistry.java]: Comment above seems to be out-of-date now.
I guess the bigger issue is I don't understand how/if this deals with paging and the identified problem that we may need to page through a large number of tags in order to get images.

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- dockstore-common/src/main/java/io/dockstore/common/Registry.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dockstore/dockstore/pullRequest?prid=5105070&bid=16789291)

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/QuayImageRegistry.java (hunk 8 above)
Comment (NatalieEO) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/QuayImageRegistry.java]: yeah, that is in another ticket https://ucsc-cgl.atlassian.net/browse/DOCK-1211

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- dockstore-common/src/main/java/io/dockstore/common/Registry.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dockstore/dockstore/pullRequest?prid=5105070&bid=16789291)

**Comment by NatalieEO** [comment]
> Looks like the DockstoreTool model is being changed in Swagger, are there breaking changes for the frontend?
😢 yes there are

**Pull request merged**